README.md
# Local Gemstash for Ruby Development in Docker Containers [![Available on Docker Hub](https://img.shields.io/docker/pulls/ianfixes/local_gemstasher.svg)](https://hub.docker.com/r/ianfixes/local_gemstasher) [![Docker Image version](https://img.shields.io/docker/v/ianfixes/local_gemstasher.svg)](https://hub.docker.com/r/ianfixes/local_gemstasher)

**Problem**: Specifying your gem dependencies by `path:` won't work if you are in a `docker build` environment unless you add your entire development directory to the build context.

**Solution**: A local gemstash server that runs in a docker container, monitoring your local gems for changes and automatically updating them on the local server when they do.


Presented here in hopes that it can be further refined by the rubygems.org team.

## Background

Let's say that you have a mix of public and private gems in development:

* `/path/to/development/corporate/private_gem`
* `/path/to/development/public/public_gem`
* `/path/to/development/corporate/private_project`

Ordinarily, your `/path/to/development/corporate/private_project/Gemfile` might say:

```ruby
source 'https://rubygems.org'

gem 'public_gem', '=1.1.1'

source 'https://private-gem-repo.org' do
  gem 'private_gem', '=2.2.2'
end
```

But during development, you might be working with unreleased versions of the dependencies.  You might change your `Gemfile` like this:

```ruby
source 'https://rubygems.org'

#gem 'public_gem', '=1.1.1'
gem 'public_gem', path: '../../public/public_gem'

source 'https://private-gem-repo.org' do
  #gem 'private_gem', '=2.2.2'
  gem 'private_gem', path: '../private_gem'
end
```

Now assume you want to test that application using Docker.  Your `/path/to/development/corporate/private_project/Dockerfile` might say:

```Dockerfile
FROM ruby:2.6-alpine

COPY Gemfile .
RUN gem install bundler && bundle install
```

but `docker build .` will fail because `../../public/public_gem` and `../private_gem` won't be in the build context.  e.g.:

```
 ---> Running in b6a85624a938
Successfully installed bundler-2.2.13
1 gem installed
The path `/public/public_gem` does not exist.
The command '/bin/sh -c gem install bundler && bundle install' returned a non-zero code: 13
```

Let's start the `local-gemstasher` server to fix that.

## Building the server image locally

We'll call it `local_gemstasher`.

```console
$ docker build -t local_gemstasher .
```


## Downloading the pre-built image

This image is available [on DockerHub]((https://hub.docker.com/r/ianfixes/local_gemstasher)), so without installing anything from github you can simply

```console
$ docker pull ianfixes/local_gemstasher
```

## Starting the server

Let's assume that you do your development in a directory called `/path/to/development`, and it conains the following:

```
.
├── _public
│   └──_public_gem
│      ├── public_gem.gemspec
│      └── ...
├── _corporate
│   └──_private_gem
│      ├── private_gem.gemspec
│      └── ...
```

We will need to map this directory to `/gems` in the container (which we can safely do as read-only).  We will then need to list the gem directories, relative to that base, that we want to monitor for changes.  In this case, those paths would be `public/public_gem` and `corporate/private_gem`.

The full invocation of `docker run` for our example is as follows:

```console
$ docker run --rm                        \
  --publish 9293:9292                    \
  --volume /path/to/development:/gems:ro \
  ianfixes/local_gemstasher              \
    public/public_gem                    \
    corporate/private_gem
```

This will launch the server on `http://localhost:9293`.  It will immediately find the available gems, build them, and push them to the local server.

It will also watch these gems for any changes (at the filesystem level) and **automatically rebuild and re-push them as necessary**.  (Note that this can involve a reboot of the internal server, because `gemstash` does not allow overwriting nor re-pushing a yanked gem version.  Such is life.)

## Using the server from Docker builds

No changes are required to your `Dockerfile`.

You will need to tell your `Gemfile` to use the mirrored server as a source.  Note that `host.docker.internal` is the server hostname from the perspective of a Docker build:

```ruby
source 'https://rubygems.org'

# wrap the old lines in the docker server source
# you may need to adjust version numbers as appropriate
source 'http://host.docker.internal:9293/private' do
  gem 'public_gem', '=1.1.1'
  gem 'private_gem', '=2.2.2'
end
```

### Gotchas

Note that when you `bundle install`, you will have cached the version of `public_gem` and `private_gem` that were available on the local_gemstash server at the time -- bundler will make no attempt to re-download them, and there doesn't seem to be an option to `--redownload` an individual gem in bundler.

If you are using local_gemstasher to support using local gems in a `docker build`, this should be no problem -- assuming that you have already worked around the issue of docker layer caching in some way (up to and including the `--no-cache` flag for `docker build`).

However, if your use case involves also installing to your local machine, you will need to make use of the following:

```console
$ gem uninstall --force --executables --ignore-dependencies <name of gem> \
  && bundle install
  && <whatever command you wanted to run>
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- colorized text
### Changed

### Fixed
- pre-release gems are now properly handled, fixing a bug where refreshing them would just cause a server error

### Deprecated

### Removed

### Security


## [0.0.1] - 2021-03-05
Initial release


[Unreleased]: https://github.com/ianfixes/local-gemstasher/compare/0.0.1...HEAD
[0.0.1]: https://github.com/ianfixes/local-gemstasher/releases/tag/0.0.1


watch_and_publish.rb
require 'filewatcher'
require 'socket'
require 'timeout'
require 'shellwords'
require 'rubygems'
require 'ostruct'
require 'colorize'

def usage
  [
    "Usage:",
    "  docker run --rm -p 9292:9293 -v /path/to/gems/basedir:/gems:ro IMAGE_ID <gem1/path> [gem2/path...]",
    "",
    "This container is meant to run locally, with a volume-mounted base",
    "directory and a list of gem directories to serve, specified as their",
    "relative paths from the base directory.  It will listen on port 9293."
  ].join("\n")
end

def nope(str)
  STDERR.puts
  STDERR.puts "ERROR"
  STDERR.puts
  STDERR.puts str
  exit(1)
end

# https://stackoverflow.com/a/9017896
def port_open?(port)
  seconds = 1
  Timeout::timeout(seconds) do
    TCPSocket.new("127.0.0.1", port).close
    true
  rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH
  end
rescue Timeout::Error
end

# print a command output as annotated
def annotated_command(*args, **kwargs)
  IO.popen(*args, **kwargs, err: [:child, :out]) do |io|
    name = yield(io) # this is a backwards way to do it, but I need a way to not mangle input args here
    #STDERR.puts "#{name}$ #{args} #{kwargs}"
    io.each do |line|
      STDERR.puts "#{name}: #{line}" unless line.strip.empty?
    end
  end
end

# TODO: this writes the config file as a side effect of generating the command.
# That's because I wasted about 5 hours not knowing that specifying a config file
# to start the server (but not specifying one for the auth) means "invalid auth".
# So it works now and it's here now.
def gemstash_command(*args)
  config = Pathname.new(ENV["APPDIR"]) + "config.yml"
  unless config.exist?
    File.write(
      config,
      [
        "---",
        ":base_path: #{ENV["GEMSTASH_WORKDIR"]}",
        ":protected_fetch: false",
        # ":log_file: :stdout"    # note that this adds extra lines to the output of "gemstash authorize"
      ].join("\n")
    )
  end

  [
    ["bundle", "exec", "gemstash"],
    args,
    ["--config-file", config] # we need this EVERY TIME OR PUSHES WILL FAIL FOR NON-OBVIOUS REASONS
  ].flatten.shelljoin
end

def await_port_status(desired_state)
  loop do
    break if !!port_open?(ENV['SERVER_PORT']) == !!desired_state
    STDERR.puts("Waiting for Gemstash port_open? status to be #{desired_state}")
    sleep(1)
  end
end

# Run gemstash in a thread.  accept a block with whatever to do while we wait for the port to open
# return the thread, server IO.popen object, and the auth key to access the server
def run_gemstash
  await_port_status(false) # don't start until any old servers are dead

  STDERR.puts "Clearing gemstash dir".yellow
  annotated_command("rm -rfv #{ENV["GEMSTASH_WORKDIR"]}/*") { "SERVER CLEANUP" }

  # generate the new key
  key = `#{gemstash_command("authorize")}`.lines.first.split(":")[1].strip
  STDERR.puts "Using GEM_HOST_API_KEY=#{key}".yellow

  # generate the new thread and IO object
  io = nil
  STDERR.puts "Launching gemstash thread".yellow
  thread = Thread.new do
    annotated_command(gemstash_command("start", "--no-daemonize")) do |io_obj|
      io = io_obj
      "GEMSTASH"
    end
  end

  await_port_status(true) # don't return until the new server is ready
  OpenStruct.new(thread: thread, io: io, key: key)
end

# load info from the gemspec.  Don't Gem::Specification load in this binding
# or else it will be immutable!  explicit subprocess avoids that problem
def load_gemspec(path)
  OpenStruct.new(
    name: `ruby -e "puts Gem::Specification.load('#{path}').name"`.strip,
    version: `ruby -e "puts Gem::Specification.load('#{path}').version"`.strip,
  )
end

def server_address
  "http://localhost:#{ENV['SERVER_PORT']}/private"
end

# get the gemspec file from a directory
def gemspec_of_dir(gem_dir)
  gemspec = gem_dir.glob("*.gemspec").first
  STDERR.puts "Found #{gemspec}: #{gemspec.exist?}"
  gemspec
end

# get gem information from a directory that contains it
def gem_info(gem_dir)
  gemspec = gemspec_of_dir(gem_dir)
  spec = load_gemspec(gemspec)
  STDERR.puts "It describes name=#{spec.name} version=#{spec.version}"
  [spec.name, spec.version]
end

# ask our local server if it has our gem
def gem_version_exists(name, version)
  result = `gem list -r --clear-sources --source #{server_address} --all --prerelease -e #{name}`
  result.lines.each { |l| STDERR.puts "GEM LIST: #{l}" }
  lines = result.lines.select { |l| l.start_with?(name) }
  return false if lines.empty?

  lines.first.match(/\(([^)]+)\)/)[1].split(", ").include?(version)
end

# gem existence wrapper but by directory instead
def gem_version_exists_by_dir(gem_dir)
  name, version = gem_info(gem_dir)
  gem_version_exists(name, version)
end

# push gem to local server
def publish_gem(gem_dir, key)
  gemspec = gem_dir.glob("*.gemspec").first
  name, version = gem_info(gem_dir)

  tmp_build_gem = Pathname.new(ENV["GEM_BUILD_DIR"]) + "tmp.gem"
  tmp_build_gem.unlink if tmp_build_gem.exist?
  return STDERR.puts "FAILED TO DELETE OLD TMP GEM".red if tmp_build_gem.exist?

  Dir.chdir(gem_dir) do
    annotated_command(["gem", "build", "-o", tmp_build_gem.to_s, gemspec.to_s].shelljoin) { "GEM BUILD" }
  end
  return STDERR.puts "FAILED TO BUILD #{gemspec}".red unless tmp_build_gem.exist?

  annotated_command(
    {"GEM_HOST_API_KEY" => key},
    ["gem", "push", "--host", server_address, tmp_build_gem.to_s].shelljoin
  ) { "GEM PUSH" }

  completion_message = "Gem pushed successfully"
  completed_result = gem_version_exists(name, version) ? "#{name} (#{version})".green : "no".red
  STDERR.puts "#{completion_message.yellow}: #{completed_result}"
rescue Exception => e
  STDERR.puts "publish_gem(#{gem_dir}) FAILED: #{e}".red
end


# Get the list of gems requested by the user and as a side effect print some
# warnings about whether the directories exist.  Also, convert them to full
# paths here because threads and Dir.chdir don't mix (unless you like warnings)
def specified_gems
  gem_basedir = ENV["LOCAL_GEMS_DIR"]
  STDERR.puts "Will look for gems relative to local mount volume #{gem_basedir}".yellow
  Dir.chdir(gem_basedir) do
    gem_basedir_contents = Dir["*"]

    if gem_basedir_contents.size == 1 && gem_basedir_contents.first == ENV["EMPTINESS_CHECK"]
      nope("It appears that no gem directory was mounted.\n#{usage}")
    end

    ARGV.map { |d| Pathname.new(gem_basedir) + (d.end_with?("/") ? d : "#{d}/") }.each do |gemdir|
      what = "WARNING: requested gem dir '#{gemdir}'"
      next STDERR.puts "#{what} doesn't exist (yet?)".red unless gemdir.exist?
      next STDERR.puts "#{what} is a file".red if gemdir.file?
      next STDERR.puts "#{what} isn't a dir ???".red unless gemdir.directory?
    end
  end

end

begin
  nope("No gem paths were specified\n#{usage}".red) if ARGV.empty?

  # read inputs and print warnings as appropriate
  gems = specified_gems

  stash_runner = run_gemstash

  Dir.chdir(ENV["LOCAL_GEMS_DIR"]) do
    Filewatcher.new(gems, interval: 0.7, immediate: true).watch do |changes|
      STDERR.puts
      valid_gems = gems.select(&:exist?).map(&:realpath) # because maybe they were added

      # find out which gems were changed. on the first run, ALL of them are changed
      changed_gems = []
      changes.each do |filename, event|
        if filename.empty? && event.empty?  # "immediate" -- first time
          changed_gems = valid_gems
        else
          Pathname.new(filename).ascend { |v| changed_gems << v if valid_gems.include?(v) }
        end
      end
      changed_gems = changed_gems.uniq
      STDERR.puts "Reacting to changes in gems: #{changed_gems.map(&:to_s)}".yellow

      # can't push existing gems.  can't re-push yanked gems. can only reboot the server with blank config.
      if changed_gems.any? { |gem_dir| gem_version_exists_by_dir(gem_dir) }
        STDERR.puts "Rebooting server".yellow
        Process.kill("KILL", stash_runner.io.pid)
        stash_runner.thread.exit
        stash_runner.thread.join
        stash_runner = run_gemstash
      end
      STDERR.puts "Publishing gems".yellow
      changed_gems.each { |gem_dir| publish_gem(gem_dir, stash_runner.key) }
    end
  end

ensure
  stash_runner.join
end
